File: icpc/World_Finals_2022/ICPC_WF2022_Q.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

const int MxN = 1000010;

struct fenwick_tree {
	ll t[MxN];

	void build(int N) {
		for(int idx = 0; idx<=N; ++idx) {
			t[idx] += 1ll;
			int nxt = idx + (idx & -idx);
			if(nxt < MxN) {
				t[nxt] += t[idx];
			}
		}
	}

	void update(int idx, ll v) {
		for(; idx<MxN; idx+=idx&-idx) {
			t[idx] += v;
		}
	}

	ll query(int idx) {
		ll res = 0ll;
		for(; idx>0; idx-=idx&-idx) {
			res += t[idx];
		}
		return res;
	}
} fw;

signed main(int argc, char *argv[]) {
	cin.tie(nullptr)->ios::sync_with_stdio(false);
	int n;
	cin >> n;
	fw.build(n);
	ll answer = 2ll * n, x;
	for(int i=1, last = n + 1; i<=n; ++i) {
		cin >> x;
		fw.update(x, -1);
		answer += fw.query(n) - fw.query(min((int) x, last));
		last = x;
	}
	cout << fixed << setprecision(4) << ((double) answer) / 2.00 << "\n";
	return 0;
}

File: icpc/World_Finals_2022/ICPC_WF2022_Q.py
class fenwick:
    def build(self, n):
        i = 1
        while i <= n:
            self.t[i] += 1;
            nxt = i + (i &-i);
            if nxt < len(self.t):
                self.t[nxt] += self.t[i]
            i = i + 1

    def update(self, idx, v):
        while idx < len(self.t):
            self.t[idx] += v
            idx = idx + (idx & -idx)

    def query(self, idx):
        res = 0
        while idx > 0:
            res += self.t[idx]
            idx = idx - (idx & -idx)
        return res

    def __init__(self, n):
        self.t = [0 for i in range(0, n + 1)]

n = int(input())
fw = fenwick(n)
fw.build(n)
answer = 2 * n
last = n + 1

for x in map(int, input().split()):
    fw.update(x, -1)
    answer += fw.query(n) - fw.query(min(last, x))
    last = x

print(f"{answer / 2:.4f}")
